power_grid_model_c/power_grid_model/include/power_grid_model/math_solver/common_solver_functions.hpp

```diff
@@ -57,22 +59,94 @@ template <symmetry_tag sym> inline void copy_y_bus(YBus<sym> const& y_bus, Compl
+// The current injection of source i to the bus in 012 domain is:
+// i_inj_0 = y_ref_i_0 / y_ref_t_0 * i_inj_t_0
+// i_inj_1 = y_ref_i_1 * [(u_ref_i_1 - i_ref_t_1 / y_ref_t_1) + i_inj_t_1 / y_ref_t_1]
+// i_inj_2 = y_ref_i_2 / y_ref_t_2 * i_inj_t_2
+// Note: u_ref_i_0 = u_ref_i_2 = 0
+// 012 domain is used to solve the numerical instability as for the symmetric case.
+inline void calculate_multiple_source_result(IdxRange const& sources, YBus<asymmetric_t> const& y_bus,
+                                             PowerFlowInput<asymmetric_t> const& input,
+                                             ComplexValue<asymmetric_t> const& i_inj_t,
+                                             SolverOutput<asymmetric_t>& output, Idx const& bus_number) {
+    std::vector<std::pair<power_grid_model::DoubleComplex, power_grid_model::DoubleComplex>> const y0_y1 =
+        y_bus.math_model_param().source_param_y0_y1;
+    ComplexValue<asymmetric_t> const y_ref_t_012 =
+        std::accumulate(sources.begin(), sources.end(), ComplexValue<asymmetric_t>{},
+                        [&](ComplexValue<asymmetric_t> sum, Idx const source) {
+                            sum(0) += y0_y1[source].first;
+                            sum(1) += y0_y1[source].second;
+                            sum(2) = sum(1); // y1 = y2
+                            return sum;
+                        });
+    DoubleComplex const i_ref_1_t =
+        std::accumulate(sources.begin(), sources.end(), DoubleComplex{}, [&](DoubleComplex sum, Idx const source) {
+            return input.source[source] * y0_y1[source].second + sum;
+        });
+    ComplexValue<asymmetric_t> const i_inj_t_012 = dot(get_sym_matrix_inv(), i_inj_t);
+
     for (Idx const source : sources) {
-        ComplexValue<sym> const u_ref{input.source[source]};
-        ComplexTensor<sym> const y_ref = y_bus.math_model_param().source_param[source];
-        output.source[source].i = dot(y_ref, u_ref - output.u[bus_number]);
+        DoubleComplex const y_ref_i_1_over_y_ref_t_1 = y0_y1[source].second / y_ref_t_012[1];
+        DoubleComplex const i_inj_i_1_lhs =
+            y_ref_i_1_over_y_ref_t_1 * (input.source[source] * y_ref_t_012[1] - i_ref_1_t);
+        DoubleComplex const i_inj_i_1 = i_inj_i_1_lhs + (y_ref_i_1_over_y_ref_t_1 * i_inj_t_012(1));
+
+        DoubleComplex const i_inj_i_0 = (y0_y1[source].first / y_ref_t_012[0]) * i_inj_t_012(0);
+        DoubleComplex const i_inj_i_2 = (y0_y1[source].second / y_ref_t_012[2]) * i_inj_t_012(2);
+        ComplexValue<asymmetric_t> const i_inj_012{i_inj_i_0, i_inj_i_1, i_inj_i_2};
+        output.source[source].i = dot(get_sym_matrix(), i_inj_012);
         output.source[source].s = output.u[bus_number] * conj(output.source[source].i);
     }
 }
 
+// Current implementation avoids numerical instability. For details refer to:
+// https://github.com/PowerGridModel/power-grid-model-workshop/blob/experiment/source-calculation/source_calculation/source_calculation.ipynb
```

Comment:
Should we make this notebook official and provide a more proper linkage here?

Reply:
@TonyXiang8787 What do you think?

Reply:
The notebook is already tagged so the everyone can read it.